### PR TITLE
Finish model compression for se_r descriptor!

### PR DESCRIPTION
**Pull request by liangadam**
Now model compression is available for se_r descriptor.

The compressed result has been checked with lammps , both on CPU and GPU. It's correct and the deviation is within 1e-8 in 1000 steps.

Acceleration performance: CPU 300%, GPU 120% (may be due to the small test system).

### CODE REVIEW

File: deepmd/descriptor/se_a.py [hunk 1]
```diff
@@ -120,8 +120,6 @@ def __init__ (self,
         """
         Constructor
         """
-        if rcut < rcut_smth:
-            raise RuntimeError("rcut_smth (%f) should be no more than rcut (%f)!" % (rcut_smth, rcut))
```
Comment (njzjz) [inline deepmd/descriptor/se_a.py:124]: Why did you delete these two lines?

File: deepmd/descriptor/se_r.py [hunk 2]
```diff
@@ -83,8 +85,6 @@ def __init__ (self,
         #        .add("activation_function", str, default = "tanh") \
         #        .add("precision",           str, default = "default")
         # class_data = args.parse(jdata)
-        if rcut < rcut_smth:
-            raise RuntimeError("rcut_smth (%f) should be no more than rcut (%f)!" % (rcut_smth, rcut))
```
Comment (njzjz) [inline deepmd/descriptor/se_r.py:87]: Same here

File: deepmd/utils/tabulate.py [hunk 3]
```diff
@@ -82,20 +82,31 @@ def __init__(self,
...
-            self.sel_a = self.graph.get_operation_by_name('DescrptSeA').get_attr('sel_a')
-            self.prod_env_mat_op = self.graph.get_operation_by_name ('DescrptSeA')
+        if isinstance(self.descrpt, deepmd.descriptor.DescrptSeR):
+            try:
+                self.sel_a = self.graph.get_operation_by_name('ProdEnvMatR').get_attr('sel')
+                self.prod_env_mat_op = self.graph.get_operation_by_name ('ProdEnvMatR')
+            except Exception:
+                self.sel_a = self.graph.get_operation_by_name('DescrptSeR').get_attr('sel')
+                self.prod_env_mat_op = self.graph.get_operation_by_name ('DescrptSeR')
+        else:
```
Comment (wanghan-iapcm) [inline deepmd/utils/tabulate.py:92]: `elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeA):` is recommended. 

And the conditions stop when `else` which capture the case self.descrpt is neither DescrptSeA nor DescrptSeR, and throw an error.

File: deepmd/utils/tabulate.py [hunk 4]
```diff
@@ -82,20 +82,31 @@ def __init__(self,
...
-            self.sel_a = self.graph.get_operation_by_name('ProdEnvMatA').get_attr('sel_a')
-            self.prod_env_mat_op = self.graph.get_operation_by_name ('ProdEnvMatA')
-        except Exception:
-            self.sel_a = self.graph.get_operation_by_name('DescrptSeA').get_attr('sel_a')
-            self.prod_env_mat_op = self.graph.get_operation_by_name ('DescrptSeA')
+        if isinstance(self.descrpt, deepmd.descriptor.DescrptSeR):
+            try:
+                self.sel_a = self.graph.get_operation_by_name('ProdEnvMatR').get_attr('sel')
+                self.prod_env_mat_op = self.graph.get_operation_by_name ('ProdEnvMatR')
+            except Exception:
```
Comment (njzjz) [inline deepmd/utils/tabulate.py:89]: Use `KeyError` instead of `Exception`. https://www.tensorflow.org/api_docs/python/tf/Graph#get_operation_by_name gives what `get_operation_by_name` raises.

File: deepmd/descriptor/se_r.py [hunk 5]
```diff
@@ -229,6 +230,60 @@ def compute_input_stats (self,
...
+        suffix : str, optional
+                The suffix of the scope
+        """
+        assert (
+            not self.filter_resnet_dt
+        ), "Model compression error: descriptor resnet_dt must be false!"
+
+        for ii in range(len(self.filter_neuron) - 1):
+            if self.filter_neuron[ii] * 2 != self.filter_neuron[ii + 1]:
+                raise RecursionError(
```
Comment (njzjz) [inline deepmd/descriptor/se_r.py:268]: Why is it `RecursionError`? I think it should be `NotImplementedError` instead.

File: deepmd/utils/tabulate.py (hunk 3 above)
Comment (liangadam) [reply deepmd/utils/tabulate.py:92]: Yes, it‘s safer. I will change it~

File: deepmd/descriptor/se_r.py (hunk 2 above)
Comment (liangadam) [reply deepmd/descriptor/se_r.py:87]: Sorry for that. This may be because I used the previous version of code when committing mine. I will recall it.

File: deepmd/utils/tabulate.py (hunk 4 above)
Comment (liangadam) [reply deepmd/utils/tabulate.py:89]: OK~

File: deepmd/descriptor/se_r.py (hunk 5 above)
Comment (liangadam) [reply deepmd/descriptor/se_r.py:268]: OK~

**Comment by njzjz** [comment]
Could you also update https://docs.deepmodeling.org/projects/deepmd/en/latest/freeze/compress.html？

**Comment by liangadam** [comment]
> Could you also update https://docs.deepmodeling.org/projects/deepmd/en/latest/freeze/compress.html?

Sure~

File: deepmd/utils/tabulate.py [hunk 6]
```diff
@@ -248,6 +298,18 @@ def _get_matrix(self):
...
+                    for ii in range(0, self.ntypes):
+                        node = self.embedding_net_nodes[f"filter_type_all{self.suffix}/matrix_{layer}_{ii}"]
+                        matrix["layer_" + str(layer)].append(tf.make_ndarray(node))
+                else:
+                    for ii in range(0, self.ntypes * self.ntypes):
+                        if (ii // self.ntypes, ii % self.ntypes) not in self.exclude_types:
+                            node = self.embedding_net_nodes[f"filter_type_{ii // self.ntypes}{self.suffix}/matrix_{layer}_{ii % self.ntypes}"]
+                            matrix["layer_" + str(layer)].append(tf.make_ndarray(node))
+                        else:
+                            matrix["layer_" + str(layer)].append(np.array([]))
```
Comment (amcadmus) [inline deepmd/utils/tabulate.py:312]: Please add an `else` to raise the error of un-supported descritpor

File: deepmd/utils/tabulate.py [hunk 7]
```diff
@@ -225,6 +263,18 @@ def _get_bias(self):
...
+                    for ii in range(0, self.ntypes):
+                        node = self.embedding_net_nodes[f"filter_type_all{self.suffix}/bias_{layer}_{ii}"]
+                        bias["layer_" + str(layer)].append(tf.make_ndarray(node))
+                else:
+                    for ii in range(0, self.ntypes * self.ntypes):
+                        if (ii // self.ntypes, ii % self.ntypes) not in self.exclude_types:
+                            node = self.embedding_net_nodes[f"filter_type_{ii // self.ntypes}{self.suffix}/bias_{layer}_{ii % self.ntypes}"]
+                            bias["layer_" + str(layer)].append(tf.make_ndarray(node))
+                        else:
+                            bias["layer_" + str(layer)].append(np.array([]))
```
Comment (amcadmus) [inline deepmd/utils/tabulate.py:277]: Please add an `else` to raise the error of un-supported descritpor

File: deepmd/utils/tabulate.py [hunk 8]
```diff
@@ -317,6 +379,9 @@ def _get_env_mat_range(self,
             var = np.square(sw / (min_nbor_dist * self.dstd[:, 1:4]))
             lower = np.min(-var)
             upper = np.max(var)
+        elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeR):
+            lower = np.min(-self.davg[:, 0] / self.dstd[:, 0])
+            upper = np.max(((1 / min_nbor_dist) * sw - self.davg[:, 0]) / self.dstd[:, 0])
```
Comment (amcadmus) [inline deepmd/utils/tabulate.py:384]: Please add an `else` to raise the error of un-supported descritpor

File: deepmd/utils/tabulate.py [hunk 9]
```diff
@@ -342,6 +407,10 @@ def _get_layer_size(self):
                 layer_size = len(self.embedding_net_nodes) // (self.ntypes * 2)
         elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeT):
             layer_size = len(self.embedding_net_nodes) // int(comb(self.ntypes + 1, 2) * 2)
+        elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeR):
+            layer_size = len(self.embedding_net_nodes) // ((self.ntypes * self.ntypes - len(self.exclude_types)) * 2)
+            if self.type_one_side :
+                layer_size = len(self.embedding_net_nodes) // (self.ntypes * 2)
         return layer_size
```
Comment (amcadmus) [inline deepmd/utils/tabulate.py:414]: Please add an `else` to raise the error of un-supported descritpor

File: deepmd/utils/tabulate.py [hunk 10]
```diff
@@ -352,6 +421,10 @@ def _get_table_size(self):
                 table_size = self.ntypes
         elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeT):
             table_size = int(comb(self.ntypes + 1, 2))
+        elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeR):
+            table_size = self.ntypes * self.ntypes
+            if self.type_one_side :
+                table_size = self.ntypes
```
Comment (amcadmus) [inline deepmd/utils/tabulate.py:427]: Please add an `else` to raise the error of un-supported descritpor

**Review by amcadmus — CHANGES_REQUESTED** [review]
The unit test is missing. Please add them to check if the compressed OP gives the same result as the original OP. 

You may take `source/lib/tests/test_tabulate.cc` as an example.

**Review by njzjz — APPROVED** [review]

File: deepmd/utils/tabulate.py [hunk 11]
```diff
@@ -82,20 +82,40 @@ def __init__(self,
...
+                self.sel_a = self.graph.get_operation_by_name('DescrptSeR').get_attr('sel')
+                self.prod_env_mat_op = self.graph.get_operation_by_name ('DescrptSeR')
+        elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeA):
+            try:
+                self.sel_a = self.graph.get_operation_by_name('ProdEnvMatA').get_attr('sel_a')
+                self.prod_env_mat_op = self.graph.get_operation_by_name ('ProdEnvMatA')
+            except KeyError:
+                self.sel_a = self.graph.get_operation_by_name('DescrptSeA').get_attr('sel_a')
+                self.prod_env_mat_op = self.graph.get_operation_by_name ('DescrptSeA')
+        elif isinstance(self.descrpt, deepmd.descriptor.DescrptSeT):
```
Comment (njzjz) [inline deepmd/utils/tabulate.py:99]: I think `DescrptSeA` and `DescrptSeT` are the same here?

Comment (liangadam) [reply deepmd/utils/tabulate.py:99]: Yes, they are the same. But considering the possible changes later, I have explicitly distinguished them.

File: deepmd/utils/tabulate.py (hunk 10 above)
Comment (liangadam) [reply deepmd/utils/tabulate.py:427]: OK~

**Comment by liangadam** [comment]
> The unit test is missing. Please add them to check if the compressed OP gives the same result as the original OP.
> 
> You may take `source/lib/tests/test_tabulate.cc` as an example.

OK. I will add an UT for this.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
It is very nice to see the new python tests checking the output of the original model is identical to the compressed model. 

Could you add the unittest for the new methods `tabulate_fusion_se_r_*_cpu` and `tabulate_fusion_se_r_*_gpu_cuda` added in `source/lib/include/tabulate.h` by this PR?

 You may take the tests of  `tabulate_fusion_se_a_*_cpu` and `tabulate_fusion_se_a_*_gpu_cuda`  in `source/lib/tests/test_tabulate.cc`

**Comment by liangadam** [comment]
> 

OK. I will add them.

**Comment by wanghan-iapcm** [comment]
Your UT failed. Please check.

**Comment by liangadam** [comment]
https://github.com/deepmodeling/deepmd-kit/pull/1391/files#r781755472

A mistake in #1391 caused this. It couldn't pass UT but it was merged into devel.

**Review by wanghan-iapcm — APPROVED** [review]
